pennylane_lightning/core/src/utils/NDPermuter.hpp

```diff
@@ -0,0 +1,268 @@
+        // ordering).
+        std::vector<size_t> old_counter(num_indices, 0);
+        // offset is important when doing this in blocks, as it's indeed
+        // implemented.
+        std::size_t offset = 0;
+        // internal_po keeps track of interations within a block.
+        // Blocks have size `blocksize`.
+        std::size_t internal_po = 0;
+
+        T *data = data_out.data();
+        const T *scratch =
+            data_.data(); // internal pointer offers better performance than
+                          // pointer from argument
+
+        std::size_t effective_max;
+
+        while (true) {
+            // If end of entire opration, break.
+            if (po == total_dim - 1)
+                break;
+
+            internal_po = 0;
+            // Each iteration of the while block goes through a new position.
+            // Inside the while, j takes care of increasing indices properly.
+            while (true) {
+                po = 0;
+                pn = 0;
+                for (size_t i = 0; i < num_indices; i++) {
+                    po += old_super_dimensions[i] * old_counter[i];
+                    pn += new_super_dimensions[map_old_to_new_idxpos[i]] *
+                          old_counter[i];
+                }
+                small_map_old_to_new_position[po - offset] = pn;
+
+                bool complete{true};
+                for (size_t j = num_indices; j--;) {
+                    if (++old_counter[j] < shape[j]) {
+                        complete = false;
+                        break;
+                    } else {
+                        old_counter[j] = 0;
+                    }
+                }
+                // If end of block or end of entire operation, break.
+                if ((++internal_po == blocksize_) || (po == total_dim - 1))
+                    break;
+                // If last index (0) was increased, then go back to fastest
+                // index.
+                if (complete)
+                    break;
+            }
+            // Copy data for this block, taking into account offset of
+            // small_map...
+            effective_max = std::min(blocksize_, remaining_data);
+            for (size_t p = 0; p < effective_max; p++) {
+                data[small_map_old_to_new_position[p]] = scratch[offset + p];
+            }
+
+            offset += blocksize_;
+            remaining_data -= blocksize_;
+        }
+    }
+
+    /**
+     * @brief Return-based transpose operation. See `Permuter` class for more
+     * details.
+     */
+    template <class T>
+    std::vector<T> Transpose(std::vector<T> data_,
+                             const std::vector<size_t> &shape,
+                             const std::vector<std::string> &old_indices,
+                             const std::vector<std::string> &new_indices) {
+        std::vector<T> data_out(std::move(data_));
+        Transpose(data_, shape, data_out, old_indices, new_indices);
```

Comment:
```suggestion
        std::vector<T> data_out(std::move(data_in));
        Transpose(data_in, shape, data_out, old_indices, new_indices);
```

Reply:
To be resolved by later importing directly from JET. Resolving.